src/analyze/analyze.c

```diff
@@ -491,11 +491,40 @@ static int pretty_boot_time(sd_bus *bus, char **_buf) {
         size_t size;
         char *ptr;
         int r;
+        usec_t activated_time = USEC_INFINITY;
+        _cleanup_free_ char* path = NULL, *unit_id = NULL;
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
 
         r = acquire_boot_times(bus, &t);
         if (r < 0)
                 return r;
 
+        path = unit_dbus_path_from_name(SPECIAL_DEFAULT_TARGET);
```

Comment:
`SPECIAL_DEFAULT_TARGET` doesn't seem to be a reliable way to get the target `systemd` has reached. For example, `systemd.unit` can be used to specify a different target. When this happens, `systemd-analyze` prints quite strange numbers like the following:
```sh
-bash-4.3# systemctl get-default
testsuite.target

-bash-4.3# sed 's/\x0/\n/g' /proc/1/cmdline
/lib/systemd/systemd
systemd.unit=multi-user.target

-bash-4.3# systemd-analyze
Startup finished in 831ms (userspace) = 4d 6h 27min 19.506s
testsuite.target reached after 584542y 1w 5d 13h 34min 30.877s in userspace
```
It's also possible to confuse `systemd-analyze` by changing the default target with `systemctl set-default`.

Would it be better to somehow record what target `systemd` has reached and make `systemd-analyze` get it ?

Reply:
Hmm, I thought I tested that.... I'll have to look at your case, because I thought I made sure that if default.target is not started, the second line is never printed...

right now I check that systemd returns ActiveEntreTimestampMonotonic with no error, and that the returned time is not USEC_INFINITY

If you manage to pass that condition with an unstarted unit... there might be a bug elsewhere in systemd. 

I could check that testsuite.target is started, but if you stop it manually, systemd-analyze would be incorrect again.

Anyway, i'll have a look and see what I can do ...

Reply:
Ok, I'm not able to reproduce your bug. I set my default to graphical.target and booted with systemd.unit=multi-user.target

systemd-analyze correctly reported the boot time without reporting the time to reach default.target (since it correctly analyzed that default.target was never reached)

so I was not able to reproduce the actual "bug" part of your comment (the 500k time reported)

can you give me a bit more on how you did that so I can reproduce ?

As for the second part of your comment, there is currently no easy way to know the target of the original boot after bootup, at least I didn't find it in the systemd dbus API. Maybe that could be added, but that would be a new RFE...

arguably, silently removing the second line instead of printing an error message might not be the best way to deal with unknown/unstarted default target, but again, this is probably better discussed in a separate  bug/RFE...

Reply:
I'm not sure what should be done to reproduce the bug, but it seems that you are getting `ActiveEnterTimestampMonotonic` equal to `USEC_INFINITY`, while I'm getting it equal to `0`.
```bash
$ busctl introspect org.freedesktop.systemd1 /org/freedesktop/systemd1/unit/graphical_2etarget | grep ActiveEnter
.ActiveEnterTimestamp               property  t         0                                        emits-change
.ActiveEnterTimestampMonotonic      property  t         0                                        emits-change
```

Reply:
okay... well 0 is also anobviously incorrect value, so I can check for both, but i'd really like to figure out where that's coming from...

Reply:
I'm wondering what part of the code can set `active_enter_timestamp` to `USEC_INFINITY`. Do you know if you have any patches that redefine the behaviour of `unit_new`, `unit_notify` or anything else?

Reply:
no, I just tested on master, but on a VM I build myself with buildroot.

I'll try to look if I messed up something in there.

Reply:
Thank you. I temporarily fixed the bug with the following patch:
```diff
diff --git a/src/analyze/analyze.c b/src/analyze/analyze.c
index d45c1dc..92ca955 100644
--- a/src/analyze/analyze.c
+++ b/src/analyze/analyze.c
@@ -541,7 +541,7 @@ static int pretty_boot_time(sd_bus *bus, char **_buf) {
         size = strpcpyf(&ptr, size, "%s (userspace) ", format_timespan(ts, sizeof(ts), t->finish_time - t->userspace_time, USEC_PER_MSEC));
         strpcpyf(&ptr, size, "= %s", format_timespan(ts, sizeof(ts), t->firmware_time + t->finish_time, USEC_PER_MSEC));

-        if (unit_id && activated_time != USEC_INFINITY)
+        if (unit_id && (activated_time > 0 && activated_time != USEC_INFINITY))
                 size = strpcpyf(&ptr, size, "\n%s reached after %s in userspace", unit_id, format_timespan(ts, sizeof(ts), activated_time - t->userspace_time, USEC_PER_MSEC));

         ptr = strdup(buf);
```
, which do roughly the same as `dual_timestamp_is_set`, but I agree that it would be great to figure out where `0` and `USEC_INFINITY` come from.

Reply:
Ok, I cleaned-up and updated my build env, and I now get the same error as you do. Not quite sure what happened, i'll do a followup PR

Reply:
fix waitint in #7764